### ISSUE TITLE
positive semi-definite Procrustes problem

### ISSUE THREAD

**Issue by fwmeng88** [theochem/procrustes#98]
This positive semi-definite  Procrustes problem is trying to 

<img src="https://latex.codecogs.com/svg.image?\underbrace{\min}_{\mathbf{P}&space;\succcurlyeq&space;0}&space;\|&space;\mathbf{B}&space;-&space;\mathbf{P}\mathbf{A}&space;\|_{F}^2" title="\underbrace{\min}_{\mathbf{P} \succcurlyeq 0} \| \mathbf{B} - \mathbf{P}\mathbf{A} \|_{F}^2" />

More details can be found at [Woodgate, K. G. (1993, December). A new algorithm for the positive semi-definite Procrustes problem. In Proceedings of 32nd IEEE Conference on Decision and Control (pp. 3596-3601). IEEE.](https://ieeexplore.ieee.org/document/325890).

This will be on the to-do list of our package, but not for this stage.

**Comment by PaulWAyers** [comment]
This would be really cool to have. It's also nice because it explains (finally!) why the symmetric Procrustes problem is useful!!!

**Comment by fwmeng88** [comment]
some more papers related with positive semi-definite Procrustes problem

1. [Kiskiras, J., & Halikias, G. D. (2007). A note on the complex semi‐definite matrix Procrustes problem. Numerical Linear Algebra with Applications, 14(6), 485-502.](https://onlinelibrary.wiley.com/doi/abs/10.1002/nla.535?casa_token=kx9hmy-20ocAAAAA:PWkTjL8xFxXdzTxOR00J1QnuQm6zDYcCrwL_nzEDBqHLR5rv6-hOiWxQZxeFLxH4ixpI381vNmWybeOr)
2. [Gillis, N., & Sharma, P. (2018). A semi-analytical approach for the positive semidefinite Procrustes problem. Linear Algebra and its Applications, 540, 112-137.](https://www.sciencedirect.com/science/article/pii/S0024379517306511?casa_token=moKGIrevSSkAAAAA:jp5d4xefdCqP1QljD3Z_N3dMapmLqRuGhJTknhUFETcMI0-CzlYLLX1vM--6gv1s5NnbYGmyHLE) 
3. [Jingjing, P., Qingwen, W., Zhenyun, P., & Zhencheng, C. (2019). Solution of symmetric positive semidefinite Procrustes problem. The Electronic Journal of Linear Algebra, 35, 543-554.](https://journals.uwyo.edu/index.php/ela/article/view/5167)
4. [Andersson, L. E., & Elfving, T. (1997). A constrained Procrustes problem. SIAM Journal on Matrix Analysis and Applications, 18(1), 124-139.](https://epubs.siam.org/doi/abs/10.1137/S0895479894277545?casa_token=RBI9dLogQuEAAAAA:boG1abqRMvJefhnqKkxGEcrqqLLzqVBI7OJOo9RAbZeY4c6yD2f6rslzKVV7FKBqpDL3P7OpheXf)
5. [Oviedo, H. F. (2019). A Spectral Gradient Projection Method for the Positive Semi-definite Procrustes Problem. arXiv preprint arXiv:1908.06497.](https://arxiv.org/abs/1908.06497) with matlab codes in https://www.mathworks.com/matlabcentral/fileexchange/64597-spectral-projected-gradient-method-for-the-positive-semi-definite-procrustes-problem

This list is built for future implementation.

**Comment by fwmeng88** [comment]
I would propose these algorithms be prioritized.

1. [Woodgate, K. G. (1993, December). A new algorithm for the positive semi-definite Procrustes problem. In Proceedings of 32nd IEEE Conference on Decision and Control (pp. 3596-3601). IEEE.](https://ieeexplore.ieee.org/document/325890)
2. [Jingjing, P., Qingwen, W., Zhenyun, P., & Zhencheng, C. (2019). Solution of symmetric positive semidefinite Procrustes problem. The Electronic Journal of Linear Algebra, 35, 543-554.](https://journals.uwyo.edu/index.php/ela/article/view/5167)
3.  [Oviedo, H. F. (2019). A Spectral Gradient Projection Method for the Positive Semi-definite Procrustes Problem. arXiv preprint arXiv:1908.06497.](https://arxiv.org/abs/1908.06497)

The proposed algorithms either come with numerical tests example or source code in Matlab. This makes the testing of the code easy.

**Comment by PaulWAyers** [comment]
Probably just implementing one algorithm here may suffice, at least at first. A related problem with an explicit solution is to find the closest semidefinite matrix to a given matrix (one matrix = I) or the closest semidefinite matrix with a given trace to a given matrix. There are analytic solutions here, and they can be used as testing. If it were possible, it would be nice to be able to solve the positive semi-definite Procrustes problem with a trace constraint, since often positive semidefinite matrices with specified (usually unit) trace show in up quantum-mechanical examples (as (reduced) density matrices).

**Comment by fwmeng88** [comment]
Thank you for the suggestions. I have changed the GSoC documentation accordingly. 

For the closest semidefinite matrix problem, do you mean something like https://nhigham.com/2021/01/26/what-is-the-nearest-positive-semidefinite-matrix/?

Also, an analog problem is the nearest correlation matrix problem, https://nhigham.com/2013/02/13/the-nearest-correlation-matrix/. I am not sure how this can be used for chemistry, but this is interesting, at least from a math perspective.

**Comment by PaulWAyers** [comment]
Yep, these references are good leads. The nearest correlation matrix is obviously interesting for machine learning.

The trace-constraint is yet another linear constraint (so similar to the correlation matrix) but has an explicit solution, see the appendix of [https://doi.org/10.1063/1.4994618](https://doi.org/10.1063/1.4994618).

In general, adding any set of linear constraint(s) on the solution is helpful (i.e., Tr[Q_k*P] = q_k) where P is positive semidefinite, Q_k is a linear operator (Hermitian in all the cases I can think of), and q_k is a float.
